docs/proposals/go-git-providers.md

```diff
@@ -0,0 +1,474 @@
+# go-git-providers
+
+## Abstract
+
+This proposal aims to create a library with the import path `github.com/fluxcd/go-git-providers`'
+(import name: `gitprovider`), which provides an abstraction layer for talking to Git providers
+like GitHub, GitLab and Bitbucket.
+
+This would become a new repository, specifically targeted at being a general-purpose Git provider
+client for multiple providers and domains.
+
+## Goals
+
+- Support multiple Git provider backends (e.g. GitHub, GitLab, Bitbucket, etc.) using the same interface
+- Support talking to multiple domains at once, including custom domains (e.g. talking to "gitlab.com" and "version.aalto.fi" from the same client)
+- Support both no authentication (for public repos), basic auth, and OAuth2 for authentication
+- Manipulating the following resources:
+  - **Organizations**: `GET`, `LIST` (both all accessible top-level orgs and sub-orgs)
+  - For a given **Organization**:
+    - **Teams**: `GET` and `LIST`
+  - **Repositories**: `GET`, `LIST` and `POST`
+    - **Teams**: `ADD` and `REMOVE`
+    - **Credentials**: `GET`, `LIST` and `POST`
+- Support sub-organizations (or "sub-groups" in GitLab) if possible
+- Pagination is automatically handled for `LIST` requests
+- Transparently can manage teams (collections of users, sub-groups in Gitlab) with varying access to repos
+
+## Non-goals
+
+- Support for features not mentioned above
+
```

Comment:
Maybe we should specify some kind of versioning pattern, e.g. go.mod and with `v1`, ... `v30` in the import path like e.g. the GitHub library does?